### ISSUE TITLE
Review

### ISSUE THREAD

**Issue by sokra** [GoogleChromeLabs/tooling.report#90]
@developit requested a review. Here you are:

# General

## Best practices?

In the mail you claimed:

> The tests show if and how a tool allows web developers to follow certain best practices in web development.

While I think this is true for some test cases, I also think that some test cases test very specific features or edge cases.
This make it difficult for developer to see what's the best practice is here.
I would recommend to split (or mark) the test cases into best practices and features.

## Plugins and Configuration

In my opinion the test results are very difficult to compare because they don't show how much plugin code + configuration is needed to pass a test.
In rollup tests I sometimes find very longy plugins, while parcel can pass some test without can configuration.
I mean any bundler can probably pass any test when you throw enough plugin/infrastructure code at it (or magic regexp replaces to work around the problem).
I would recommend to set a plugin + configuration LOC limit and/or display plugin + configuration LOC in the test results.

## Custom plugins

You have written quite a lot custom plugins for rollup.
While in some cases it makes sense to have them in your codebase (like non-js-resources/custom), in most cases developers won't do it this way.
In most cases such plugins should be published as general solutions as reusable open source plugins.
At least that is what I would consider as "best practice".

This especially relevant as some of the plugins in the test suite are kind of mocks, which only implement enough to treat the test as passed (if even?), but return fake urls, doesn't work for watch mode, have no caching, etc.
So in practice they are not usable if somebody tries to follow the configuration and plugin code in the test.
I would not consider these tests as passing in the current state, because the solution is not usable in a real codebase.

I assume it still possible to implement the solution in a usable, general-purpose fashion, but the current state doesn't prove that in my opinion.

In addition to that having a lot of custom plugin may make rollup appear a lot more complicated that it is.
The average developer might be afraid of custom plugin code needed to cover these features.

# tests

regarding webpack

## Code Splitting

Some comments and configurations about webpack claim a wrong assumption:
It says that when code is duplicated between chunks it is instantiated multiple times.
This is not true. Even when module code is duplicated, it's always only instantiated once, when it's part of the same webpack runtime (`optimization.runtimeChunk: "single"`).
In my opinion duplicating code can sometimes even perform better than avoiding duplicating at all.
It's actually a trade-off between transferring more code (less caching) vs. issuing more requests (more files).
More requests/files usually comes with less-effective compression and additional cost for http overhead (headers, etc.).
For HTTP 1.1 it can even cause additional delay due to round-trips because of connection limits.
Even if server and browser support HTTP 2, it may have to fallback to HTTP 1.1 because of some old middleboxes/proxies, but that's rare.
That's why webpack uses a heuristic to determine when code is duplicated (<30kb or reached 6 requests per import() limit) and when an additional file is created.

These tests are affected:

* multi-entry: comment
* shared-objects-between-entry-points: comment, webpack should pass this one with only `runtimeChunk: 'single'` in the configuration
* and search for `minSize`

You should remove `splitChunks.minSize: 0` from all test cases if you want to provide the best practice.

### dynamic-import

webpack has the guideline that: An `import()` should only need a single round-trip to be loaded. Multiple requests are allowed but must happen in parallel.

I would consider this as best practice and would recommend to add an additional test to cover that.

### splitting-modules

While your examples indeed doesn't work, there is a scenario that works in webpack and I believe in rollup too. (don't about parcel, you can try it)

It's used in practice by libraries like lodash-es.

* imported module must contain reexports to individual modules
* module with reexports must be side-effect-free (e. g. `package.json` must have `"sideEffects": false`)
* two different exports are used which point to different modules

Example:

``` js
// objects.js (must be side-effect-free)
export { foo } from "./foo";
export { bar } from "./bar";

// foo.js
export const foo = { name: 'foo' };

// bar.js
export const bar = { name: 'bar' };
```

### avoid-cascade

It's a trade-off. Hoisting all these asset hashes/url into the importMap/manifest can affect initial page load negatively.
webpack has gone the way that everything you put behind an `import()` should be loaded on-demand.
It still avoids cascade by limited the affected files to `O(1)`.
Yes it's 2 files affected instead of 1 file, but that's a cost I'm willing to pay for not needing to download all asset hashes/urls in the initial page load.

### avoid-hash

webpack 5 will support that, but I really don't think you need that in practice. Why would you want to hash only some of your `import()` generates javascript files?

## Importing modules

AMD is also found on some old npm modules or legacy codebases.

CommonJs <-> ESM interop is also probably worth testing.
Some edge-cases are handled differently between bundlers, but there is no clear spec for this and so nobody is wrong.

rollup handles CommonJS different than parcel and webpack, by converting it to ESM.
This has some limitations which makes sense to test here. (Conditional `require()`, circular dependencies with `require()` nested, strict mode)

There is some difference in how bundlers handle non-static `import()`, e. g. ``import(`./images/${name}.png`)``
This is also worth covering.

### CSS

I find a lot best practices not covered by these tests here.

In my opinion the best practice is the put the CSS into a separate .css file.
But not a single one, but a .css file per `import()` (or multiple for deduplication, similar to JS).

This CSS file should be loaded in parallel to the JS part.
Importing styles to get a url to the .css file isn't the best way in my opinion.

For styles in the entrypoint they need to be referenced as `<link rel="stylesheet">` in the HTML.

For styles behind `import()` a `<link rel="stylesheet">` should be added dynamically.
The stylesheet must be fully loaded when JS is executed otherwise a FOUC occurs.

For webpack the usual guideline here is that one `import()` should only need a single round-trip to be loaded (It can still have multiple requests when they happen in parallel).
This also applies to loading stylesheets.

For hashing: Best if changes in the CSS doesn't affect the hash of the JS part.

### Custom

How "easy" is it to integrate watching and caching for these custom types?

# Differences in general principles

It would be nice to highlight differences in higher principles.

Example:

In parcel and webpack chunks are container for module factories. This means chunks can be loaded without side effect of executing the modules in the chunk.

In rollup chunks are module. This means loading a chunk is equal to evaluating the modules it represents (depending on the `output.format`, assuming ESM here).

This can be used to explain some of the effects.

# More ideas

* build in watch mode
* memory caching, persistent caching
* file watching
* dev server infrastructure
* HMR
* typescript

# static-build

I would really like to see the final result, but sadly the build doesn't run on windows and I wasn't able to fix it easily.
That would be something you can avoid in long term when publishing your custom plugins for general usage as open source instead...!

# Disclaimer

All this is my personal opinion.
If it sounds a bit bad to rollup, this was unintended.
It's a great tool and my criticism is not against rollup itself, but the way the tests are implemented.
To me it seems like the authors favor rollup in general, which is not bad at all, but to be an objective test suite some adjustments need to be made.

# TODO list

At least what I would do...

* [ ] separate between features and best practices
* [ ] LOC config + plugins
* [ ] custom plugins -> publish
* [ ] code spliting: fix duplication assumption
* [ ] remove `splitChunks.minSize: 0`
* [ ] hasing cascade
* new test cases
  * [ ] `import()` should load all JS in parallel
  * [ ] splitting modules with `sideEffects: false`
  * [ ] CommonJS <-> ESM interop
  * [ ] dynamic `import()`s
  * [ ] condition `require()`
  * [ ] circular dependencies in CommonJS by nesting `require()`
  * [ ] non-strict mode CommonJS
  * [ ] CSS: load external stylesheet in entrypoints
  * [ ] CSS: avoid FOUC by having the stylesheet in the HTML
  * [ ] CSS: load external stylesheet in on-demand modules
  * [ ] CSS: avoid FOUC by ensuring that JS executes after CSS has loaded
  * [ ] `import()` should load all JS + CSS in parallel
  * [ ] CSS hash: changing CSS should not affect JS
  * [ ] Custom assets are watched during watch mode
  * [ ] Custom assets are cached during watch mode
* [ ] fix static-build for windows
